File: go-tui/tasks_repository.go
package main

import (
	"context"

	"github.com/getditto/ditto-go-sdk/v5/ditto"
	"github.com/google/uuid"
)

// Task is a single todo item as stored in the "tasks" collection.
type Task struct {
	ID      string `json:"_id"`
	Title   string `json:"title"`
	Done    bool   `json:"done"`
	Deleted bool   `json:"deleted"`
}

// TasksRepository owns everything specific to the tasks data: the sync
// subscription, the store observer that streams the current task list, and the
// task CRUD operations. It talks to Ditto directly through the instance vended
// by DittoManager.
type TasksRepository struct {
	ditto        *ditto.Ditto
	subscription *ditto.SyncSubscription
	observer     *ditto.StoreObserver
}

// NewTasksRepository creates a repository backed by the manager's Ditto
// instance.
func NewTasksRepository(manager *DittoManager) *TasksRepository {
	return &TasksRepository{ditto: manager.Ditto()}
}

// RegisterSubscription subscribes this peer to the tasks collection so Ditto
// syncs matching documents from other peers.
func (r *TasksRepository) RegisterSubscription() error {
	subscription, err := r.ditto.Sync().RegisterSubscription("SELECT * FROM tasks")
	if err != nil {
		return err
	}
	r.subscription = subscription
	return nil
}

// ObserveTasks registers a store observer against the local database and
// returns a channel that receives the current task list whenever it changes.
func (r *TasksRepository) ObserveTasks(ctx context.Context) (<-chan []Task, error) {
	tasksChan := make(chan []Task, 1) // Buffer size 1 - latest update wins

	observer, err := r.ditto.Store().RegisterObserver(
		"SELECT * FROM tasks WHERE NOT deleted",
		nil,
		func(result *ditto.QueryResult) {
			defer result.Close()

			tasks := parseTasks(result)
			select {
			case tasksChan <- tasks:
			case <-ctx.Done():
			}
		})
	if err != nil {
		return nil, err
	}
	r.observer = observer
	return tasksChan, nil
}

func (r *TasksRepository) CreateTask(title string) error {
	task := map[string]interface{}{
		"_id":     uuid.New().String(),
		"title":   title,
		"done":    false,
		"deleted": false,
	}

	result, err := r.ditto.Store().Execute(
		"INSERT INTO tasks DOCUMENTS (:task)",
		ditto.QueryArguments{"task": task},
	)
	if err != nil {
		return err
	}
	result.Close()
	return nil
}

func (r *TasksRepository) UpdateTask(id, title string) error {
	result, err := r.ditto.Store().Execute(
		"UPDATE tasks SET title = :title WHERE _id = :id",
		ditto.QueryArguments{
			"title": title,
			"id":    id,
		},
	)
	if err != nil {
		return err
	}
	result.Close()
	return nil
}

func (r *TasksRepository) ToggleTask(id string, done bool) error {
	result, err := r.ditto.Store().Execute(
		"UPDATE tasks SET done = :done WHERE _id = :id",
		ditto.QueryArguments{
			"done": done,
			"id":   id,
		},
	)
	if err != nil {
		return err
	}
	result.Close()
	return nil
}

func (r *TasksRepository) DeleteTask(id string) error {
	result, err := r.ditto.Store().Execute(
		"UPDATE tasks SET deleted = true WHERE _id = :id",
		ditto.QueryArguments{"id": id},
	)
	if err != nil {
		return err
	}
	result.Close()
	return nil
}

// Close cancels the observer and subscription held by the repository.
func (r *TasksRepository) Close() {
	if r.observer != nil {
		r.observer.Cancel()
	}
	if r.subscription != nil {
		r.subscription.Cancel()
	}
}

func parseTasks(result *ditto.QueryResult) []Task {
	if result == nil {
		return []Task{}
	}

	tasks := make([]Task, 0, result.ItemCount())
	for _, queryItem := range result.Items() {
		var task Task
		if err := queryItem.UnmarshalTo(&task); err != nil {
			panic(err)
		}
		tasks = append(tasks, task)
	}
	return tasks
}

File: rust-tui/src/tasks_repository.rs
use std::sync::Arc;

use anyhow::Result;
use dittolive_ditto::{store::StoreObserver, sync::SyncSubscription};
use serde::{Deserialize, Serialize};
use tokio::sync::watch;
use uuid::Uuid;

use crate::ditto_manager::DittoManager;

/// A single task document in the "tasks" collection.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Task {
    #[serde(rename = "_id")]
    pub id: String,
    pub title: String,
    pub done: bool,
    pub deleted: bool,
}

impl Task {
    pub fn new(title: String) -> Self {
        Self {
            id: Uuid::new_v4().to_string(),
            title,
            done: false,
            deleted: false,
        }
    }
}

/// Owns everything specific to the tasks data: the sync subscription, the store
/// observer that streams the current task list, and the task CRUD operations.
/// It talks to Ditto directly through the instance vended by [`DittoManager`].
///
/// The repository shares ownership of the manager via `Arc<DittoManager>`, so
/// the manager (and its persistence directory) cannot be dropped while the
/// repository is still using it.
///
/// [`DittoManager`]: crate::ditto_manager::DittoManager
pub struct TasksRepository {
    /// Ditto observer handles must be held (not dropped) to keep them alive.
    /// Observers provide the callback triggers that let us react to events.
    pub observer: Arc<StoreObserver>,

    /// Ditto subscriptions must also be held to keep them alive. Subscriptions
    /// cause Ditto to sync selected data from other peers.
    pub subscription: Arc<SyncSubscription>,

    /// Our observer sends any document updates into this watch channel.
    tasks_rx: watch::Receiver<Vec<Task>>,

    /// Shared owner of the Ditto instance. The repository reaches Ditto through
    /// [`DittoManager::ditto`]. Declared last so it drops after `observer` and
    /// `subscription` — those borrow from the Ditto instance it owns and must be
    /// torn down before it (Rust drops fields in declaration order).
    manager: Arc<DittoManager>,
}

impl TasksRepository {
    /// Register the tasks subscription and observer against the Ditto instance
    /// owned by the given manager.
    pub fn try_new(manager: Arc<DittoManager>) -> Result<Self> {
        let ditto = manager.ditto();
        let (tasks_tx, tasks_rx) = watch::channel(Vec::new());

        // Register a subscription, which determines what data syncs to this peer
        // https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
        let subscription = ditto.sync().register_subscription("SELECT * FROM tasks")?;

        // Register observer, which runs against the local database on this peer
        let observer = ditto.store().register_observer(
            "SELECT * FROM tasks WHERE NOT deleted",
            move |query_result| {
                let docs = query_result
                    .into_iter()
                    .flat_map(|it| it.deserialize_value::<Task>().ok())
                    .collect::<Vec<_>>();
                tasks_tx.send_replace(docs);
            },
        )?;

        Ok(Self {
            manager,
            tasks_rx,
            observer,
            subscription,
        })
    }

    /// The current task list, as most recently reported by the observer.
    pub fn tasks(&self) -> Vec<Task> {
        self.tasks_rx.borrow().clone()
    }

    /// Create a new task with the given title.
    pub async fn create_task(&self, title: String) -> Result<()> {
        let task = Task::new(title);
        self.manager
            .ditto()
            .store()
            .execute((
                "INSERT INTO tasks DOCUMENTS (:task)",
                serde_json::json!({ "task": task }),
            ))
            .await?;
        Ok(())
    }

    /// Set the title of the task with the given ID.
    pub async fn edit_task(&self, id: &str, title: &str) -> Result<()> {
        self.manager
            .ditto()
            .store()
            .execute((
                "UPDATE tasks SET title = :title WHERE _id = :id",
                serde_json::json!({ "title": title, "id": id }),
            ))
            .await?;
        Ok(())
    }

    /// Set the "done" flag of the task with the given ID.
    pub async fn toggle_done(&self, id: &str, done: bool) -> Result<()> {
        self.manager
            .ditto()
            .store()
            .execute((
                "UPDATE tasks SET done = :done WHERE _id = :id",
                serde_json::json!({ "id": id, "done": done }),
            ))
            .await?;
        Ok(())
    }

    /// Soft-delete the task with the given ID.
    pub async fn delete_task(&self, id: &str) -> Result<()> {
        self.manager
            .ditto()
            .store()
            .execute((
                "UPDATE tasks SET deleted = true WHERE _id = :id",
                serde_json::json!({ "id": id }),
            ))
            .await?;
        Ok(())
    }
}
